syne_tune/optimizer/transfer_learning/bounding_box.py

```diff
@@ -0,0 +1,116 @@
+        will consider a random-search on the config-space is restricted to the bounding of best evaluations of previous
+        tasks.
+        :param config_space: initial search-space to consider, will be updated to the bounding of best evaluations of
+        previous tasks
+        :param metric: objective name to optimize, must be present in transfer learning evaluations.
+        :param transfer_learning_evaluations: dictionary from task name to offline evaluations.
+        :param mode: mode to be considered, default to min.
+        :param num_hyperparameters_per_task: number of best hyperparameter to take per task when computing the bounding
+        box, default to 1.
+        """
+        self.mode = mode if mode is not None else "min"
+
+        new_config_space = self.compute_box(
+            config_space=config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            mode=self.mode,
+            num_hyperparameters_per_task=num_hyperparameters_per_task,
+            metric=metric
+        )
+        self.config_space = new_config_space
+        print(f"hyperparameter ranges of best previous configurations {new_config_space}")
+        print(f"({sp.search_space_size(new_config_space)} options)")
+        self.scheduler = scheduler_fun(new_config_space, mode, metric)
+        super(BoundingBox, self).__init__(
+            config_space=new_config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            metric_names=self.scheduler.metric_names(),
+        )
+
+    @staticmethod
+    def compute_box(
+            config_space: Dict,
+            transfer_learning_evaluations: Dict[str, TransferLearningTaskEvaluations],
+            mode: str,
+            num_hyperparameters_per_task: int,
+            metric: str
+    ) -> Dict:
+        # find the best hyperparameters on all tasks
+        best_hps = []
+        for task, evaluation in transfer_learning_evaluations.items():
+            best_hp_task_indices = evaluation.objective_values(objective_name=metric).reshape(-1).argsort()
+            if mode == 'max':
+                best_hp_task_indices = best_hp_task_indices[::-1]
+            best_hps.append(evaluation.hyperparameters.loc[best_hp_task_indices[:num_hyperparameters_per_task]])
+        hp_df = pd.concat(best_hps)
+
+        # compute bounding-box on all hyperparameters that are numerical or categorical
+        new_config_space = {}
+        for i, (name, domain) in enumerate(config_space.items()):
+            if hasattr(domain, "sample"):
+                if isinstance(domain, Categorical):
+                    hp_values = list(sorted(hp_df.loc[:, name].unique()))
+                    new_config_space[name] = sp.choice(hp_values)
+                elif hasattr(domain, "lower") and hasattr(domain, "upper"):
+                    # assume its numerical
+                    new_domain = domain
```

Comment:
This is a bit dangerous. We should better create a new instance of Domain with different upper, lower.
The Domain implementations do not assume somebody changing upper, lower after creation.

Reply:
Ok, I would update with a deep-copy.